### PR TITLE
Move external dependencies to Dependencies group in Xcode project

### PR DESCRIPTION
**Pull request by bhargavg**
- [x] Create `Dependencies` group in project
- [x] Move all external dependencies from `Source` to `Dependencies` group
- [x] Differentiate source packages & modules from external dependencies
- [x] Rebase and resolve conflicts
- [x] Refactor code
- [x] Add test cases

### CODE REVIEW

File: Sources/Xcodeproj/Module+PBXProj.swift [hunk 1]
```diff
@@ -27,16 +27,17 @@ import struct Utility.Toolchain
...
-let linkPhaseFileRefPrefix =                        "_LinkFileRef_"
-let sourceGroupFileRefPrefix =                      "__PBXFileRef_"
-let compilePhaseFileRefPrefix =                     "__src_cc_ref_"
+let rootObjectReference =                           "____RootObject_"
+let rootBuildConfigurationListReference =           "_____RootConfs_"
+let rootBuildConfigurationReference =               "_________Debug_"
+let rootGroupReference =                            "_____RootGroup_"
+let productsGroupReference =                        "______Products_"
+let sourcesGroupReference =                         "_______Sources_"
+let dependenciesGroupReference =                    "__Dependencies_"
```
Comment (mxcl) [inline Sources/Xcodeproj/Module+PBXProj.swift]: There is not a requirement to have a `_` prefix, so this diff can probably be less.

**Comment by bhargavg** [comment]
@mxcl how are we planning to test the generated `xcodeproj`? 

As @modocache mentioned in https://github.com/apple/swift-package-manager/pull/174#discussion_r55126146, are we going to build the targets to check whether everything is working? or do you have any thing in mind?

File: Sources/Xcodeproj/Module+PBXProj.swift (hunk 1 above)
Comment (bhargavg) [inline Sources/Xcodeproj/Module+PBXProj.swift]: Agreed and minimized the diff.

**Comment by mxcl** [comment]
I'm not sure there is anything really good we can do for testing here in a reasonable time, keeping in mind that Xcode itself tests for a great many configurations.

So I think what we will do is run an `xcodebuild` test for debug and release on a few Fixtures,

**Comment by bhargavg** [comment]
Squashed and rebased. Please review.

File: Sources/PackageType/Package.swift [hunk 2]
```diff
@@ -14,14 +14,16 @@ public class Package {
     public let url: String
     public let path: String
     public let name: String
+    public let isSourcePackage: Bool
     public var dependencies: [Package] = []
     public let manifest: Manifest
 
-    public init(manifest: Manifest, url: String) {
+    public init(manifest: Manifest, url: String, isSourcePackage: Bool = false) {
```
Comment (aciidb0mb3r) [inline Sources/PackageType/Package.swift]: maybe `rootPackage` makes more sense?

Comment (mxcl) [inline Sources/PackageType/Package.swift]: I think we're standardizing on the name “root package”.

Aside: I think there is a better name that can be found.

But we can correct this later.

**Comment by mxcl** [comment]
I don't want the notion of `isExternalDependency` to be part of `PackageType` this violates the responsibility boundary for this module.

Modules do not know these details about themselves. Instead this information should be passed to `Xcodeproj.generate()`.

**Comment by bhargavg** [comment]
I felt this information, ie., whether it is external package or root package, would be useful not only during Xcode project generation, but also in other places in future. That's why I've added the properties `isExternalDependency ` and `isSourcePackage` and set them during `get()` phase.

Without these properties, we have to filter through the list of packages based on path to identify whether it is root package or dependency package.

I can remove these properties and use package paths for root/dependency package identification, but I feel these properties are not violating the responsibility boundaries of these modules.

**Comment by mxcl** [comment]
> Without these properties, we have to filter through the list of packages based on path to identify whether it is root package or dependency package.

This would also be bad, I'm not suggesting we infer the information, I'm suggesting that this information be passed differently, specifically my idea is that a `Set<Module>` be returned from `transmute()` which contains the set of packages that are external and this set be passed to other modules that need it.

**Comment by mxcl** [comment]
In fact this is easier here, the module is external if it is not in the package’s `dependencies`.

**Comment by mxcl** [comment]
Alternatively, the `Package` could have a modules property. Intuitively I expect people expect this, though at this point we are only designing for ourselves.

Anyway, do which ever you like.

**Comment by bhargavg** [comment]
> specifically my idea is that a `Set<Module>` be returned from `transmute()` which contains the set of packages that are external and this set be passed to other modules that need it.

Sorry, I didn't get you. Could you please explain?

As of now `transmute()` is returning modules (root + external + test) and Products, how can we make it to return a `Set<Module>` and be able to identify which modules are root modules and which ones are external?

**Comment by mxcl** [comment]
public func transmute(packages: [Package], rootdir: String) throws -> ([Module], [Product]) {

to

    public func transmute(packages: [Package], rootPackage: Package) throws -> ([Module], [Product], externalModules: Set<Module>) {

But this might not be the best way to do it, certainly it is ugly, but it will suffice for now until I get round to refactoring it.

**Comment by bhargavg** [comment]
@mxcl Made the changes as per your previous comment. Please review.

File: Sources/Transmute/transmute().swift [hunk 3]
```diff
@@ -66,18 +66,10 @@ public func transmute(packages: [Package], rootdir: String) throws -> ([Module],
     // ensure modules depend on the modules of any dependent packages
     fillModuleGraph(packages, modulesForPackage: { map[$0]! })
 
-    var set = Set<Module>()
-    var stack = packages.flatMap{ map[$0] ?? [] }
-    var modules = [Module]()
+    let depPackages = packages.filter{ $0.path != rootdir }
```
Comment (mxcl) [inline Sources/Transmute/transmute().swift]: I'm not comfortable with this, it seems fragile. We should explicitly pass in the root package.

Comment (bhargavg) [inline Sources/Transmute/transmute().swift]: Agreed. Refactored `get()` to return source package and dependency packages separately.

**Comment by mxcl** [comment]
Great, better even than I suggested. Thanks.

**Comment by bhargavg** [comment]
Could you please run CI on this?

**Comment by goloveychuk** [comment]
@swift-ci Please test

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by bhargavg** [comment]
@mxcl, All the tests've passed. Could you please review and merge this?

**Comment by kostiakoval** [comment]
@bhargavg Conflicts

**Comment by bhargavg** [comment]
Rebased and resolved the conflicts.

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by bhargavg** [comment]
All tests passed, Could someone please merge this.

**Comment by mxcl** [comment]
Sorry, I wanted to try it out on a few packages first. Looks good, code-wise and result-wise.

**Pull request merged**